pkg/pyramid/file_tracker.go

```diff
@@ -0,0 +1,50 @@
+package pyramid
+
+import (
+	"sync"
+
+	"github.com/treeverse/lakefs/pkg/pyramid/params"
+)
+
+// fileTracker tracks file open requests in TierFS to avoid race conditions with cache rejection/eviction
+type fileTracker struct {
+	refMap map[string]int
+	mu     sync.Mutex
+}
+
+type decCallback func(int)
+
+func newFileTracker() *fileTracker {
+	return &fileTracker{
+		refMap: map[string]int{},
+		mu:     sync.Mutex{},
+	}
+}
+
+func (t *fileTracker) inc(path params.RelativePath) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	ref := 1
+
+	if val, ok := t.refMap[string(path)]; ok {
+		ref = val + 1
+		t.refMap[string(path)] = ref
+	} else {
+		t.refMap[string(path)] = ref
+	}
+}
+
+func (t *fileTracker) dec(path params.RelativePath, fn decCallback) {
```

Comment:
I think we can remove the 'fn' argument here - if dec returns true/false or count - we can defer call to 'dec' and delete the file.